### ISSUE TITLE
Menu?

### ISSUE THREAD

**Issue by dfishburn** [SidOfc/mkdx#1]
Have by chance added your mappings to a menu?
For the amount of time I spent in an .md file, I doubt I would remember the mappings.
Both Vim and gVim support menus.
The nice thing about gVim is you can tear off the menu and leave it beside your window.
I do this in several of my plugins, dbext.vim has a good example if you are interested.

**Comment by SidOfc** [comment]
Hi there!

I'm not exactly sure what you mean by "added your mappings to a menu".
Could you reference a Vim `:h` keyword so I can find out about this?
Also, could you include a screenshot of what you are looking for?

**Comment by SidOfc** [comment]
Hello again @dfishburn, perhaps you missed my last message :)
Did you find what you are looking for, or could you explain what you want in a bit more detail?

Cheers!

**Comment by SidOfc** [comment]
I'm closing this issue as it is unclear what should be done.

**Comment by dfishburn** [comment]
Sorry, have been busy to put this together.

* Vim supports user defined menus:
  * ```:h menus```
  ![image](https://user-images.githubusercontent.com/646124/34791472-14d4912e-f613-11e7-931f-3a8b71c5a78f.png)

* Vim also supports tear off menus (:h tear-off-menus).  
    * I really like this to help to learn what the standard maps your plugin offers.  
    * Since I can have it open beside the buffer I am editing.  
    * As I become a more advanced user of the plugin, I will start to remember the maps, and not need the menu.
![image](https://user-images.githubusercontent.com/646124/34791518-393b567e-f613-11e7-9b16-c03bb08a99c4.png)

Here is a simple example you can execute in your Vim right now, assuming your Vim already has a menu:
  ```noremenu &Plugin.&mkdx.Checkbox :call mkdx#ToggleCheckboxState(1)<CR>```

And the same, but supporting visual mode:
  ```vnoremenu &Plugin.&mkdx.Checkbox\ (Visual) :call mkdx#ToggleCheckboxState(1)<CR>```

And this demonstrates showing the user mapping ",my":
  ```noremenu &Plugin.&mkdx.Checkbox3<TAB>,my :call mkdx#ToggleCheckboxState(1)<CR>```

Those 3 commands create the following in Vim:
![image](https://user-images.githubusercontent.com/646124/34792072-eac7a400-f614-11e7-9be1-ba029b8a0f8b.png)


The [dbext plugin](http://www.vim.org/scripts/script.php?script_id=356) dynamically builds up the menus for normal mode, visual mode, insert mode mappings and also shows what the mapping would be.  For some of them, it even executes the mappings, as the user would using the feedkeys() Vim method.
```exec 'inoremenu <script> '.menuRoot.'.Select\ Table<TAB>'.leader.'st   <C-O>:silent call feedkeys("<C-O>'.leader.'st")<CR>'```


The code is in the plugin/dbext.vim file line 363 in the menus section.

Hope this helps.
David

**Comment by SidOfc** [comment]
Hiya!

Alright, I understand how this is supposed to look, I work in the terminal myself but this seems handy when using gvim or some other vim that supports menu's. Recently, I added a [mappings section](https://github.com/sidofc/mkdx#mappings) to the README, it is quite large indeed.

I will look into the documentation and see if I can get it working in the terminal, probably I'll ask if you can try out the branch before it is released if you're up for that (it's really simple and harmless, or well, for terminal vim anyways :P).

In any case, I will also see how it looks in macvim, I'm currently on NeoVim (tui) but will check regular Vim (tui, if menu's are supported) as well.

Cheers for the awesome examples and explanations!
Sidney Liebrand

**Comment by SidOfc** [comment]
I've added menu's like you suggested (although most are still broken in functionality, will look into it tomorrow), I am however, not entirely sure if they also show the mappings. They are there and I see them in my Vim / NeoVim menu's, but not in MacVim for instance (I do see the menu there though) some screenshots to illustrate:

**MacVim**

![image](https://user-images.githubusercontent.com/3225058/34798278-b36d8be0-f65b-11e7-90c0-10904ea76820.png)


**Vim / NeoVim**

![image](https://user-images.githubusercontent.com/3225058/34798254-9f685ff8-f65b-11e7-97ee-135d438b5a91.png)

I would like to ask if you can test the new functionality in the branch feature/menu-support.
Particularly, it would be helpful to know if you see the mappings in your menus.

For now, what you would need to do is:

$ cd ~/.vim/path/to/mkdx
$ git pull origin feature/menu-support
$ git checkout feature/menu-support

Should you at any time want to go back to master to update this plugin normally:

$ cd ~/.vim/path/to/mkdx
$ git checkout master

**Comment by SidOfc** [comment]
Hi again @dfishburn, I've tested the functionality of the menu's after some debugging and everything seems to be working now. I also noticed that the hotkeys _are_ actually displayed in MacVim but only after hovering over the menu item for a certain while, I hope they just show up like in your example screenshot.

Closing this now as this feature is indeed implemented.
If you notice anything broken still, please open a new issue.

Thank you for the great idea!
Sidney Liebrand

**Comment by dfishburn** [comment]
Sidney,

I gave this the old college try and was never able to get where I needed.
From the GitHub website, I viewed your branch and hit the download as zip.
The updated by runtimepath to include the root folder.
```
:set runtimepath+=C:\Vim\vimfiles\mkdx
:e dave.md
:echo b:did_ftplugin
1
```
No menus.

```
:echo exists('g:mkdx#map_prefix')
0
```
I believe that means no work really gets done.

So what is supposed to define and create all of your g:mkdx settings?
I do not see the autoload/mkdx.vim file creating them?

I usually do the following in my own plugins.
Create a plugin/mkdx.vim
It does some initial setup work, but leaves most of the stuff to be handled in the autoload/mkdx.vim.
If you have a look at the plugin/dbext.vim it is the one that defines all my g:dbext variables.  It does the smart thing about seeing if the user has overridden them and sets them appropriately.
I am not certain how your plugin does it's initial setup.

If you were to move most of the code from ftplugin/mkdx.vim into a function inside autoload/mkdx.vim it makes using the `:breakadd` statement easier to use and then I can run the command to create the menus again.  It is a pain having to step through line after line in the ftplugin to get to the point where it attempts to create the menu.

David

**Comment by SidOfc** [comment]
Hi David,

Can you confirm that you have the [latest version](https://github.com/SidOfc/mkdx/releases) installed by checking the version number at `:h mkdx`, it should look like this:

![image](https://user-images.githubusercontent.com/3225058/34917804-49be7892-f94b-11e7-914f-10f36860b365.png)

Also, did you indeed confirm that the plugin wasn't loaded using some other method? Like executing a mapping in a markdown file that didn't work for instance?

Another thing to note is that mkdx is a filetype plugin, as you also tested using `:echo b:did_ftplugin`. Everything is set up in the `ftplugin/markdown/mkdx.vim` file:

![image](https://user-images.githubusercontent.com/3225058/34917834-9328537c-f94b-11e7-8cf8-07f535cfb0cc.png)

This means that the menu only gets set up after loading a markdown file.
The reason you are getting `0` from `:echo exists('g:mkdx#map_prefix')` is because I deprecated the old `g:mkdx#setting_name` vars in favor of a single `g:mkdx#settings` dictionary (thus, the global vars are no longer the real "defaults"), [read more information here](https://github.com/SidOfc/mkdx#gmkdxsettings).

All the setup work gets done in `ftplugin/markdown/mkdx.vim`. This is what sets up the [plugs](https://github.com/SidOfc/mkdx/blob/2c19bb9fa9f049ee2b9e115bde097696a58b2c02/ftplugin/markdown/mkdx.vim#L56-L78), [default settings](https://github.com/SidOfc/mkdx/blob/2c19bb9fa9f049ee2b9e115bde097696a58b2c02/ftplugin/markdown/mkdx.vim#L2-L54), [mappings](https://github.com/SidOfc/mkdx/blob/2c19bb9fa9f049ee2b9e115bde097696a58b2c02/ftplugin/markdown/mkdx.vim#L140) and [menu's](https://github.com/SidOfc/mkdx/blob/master/ftplugin/markdown/mkdx.vim#L137). The mappings and menu's are built dynamically using the [`s:bindings` array](https://github.com/SidOfc/mkdx/blob/2c19bb9fa9f049ee2b9e115bde097696a58b2c02/ftplugin/markdown/mkdx.vim#L85-L116) which contains all the information to generate the correct [`exe` commands](https://github.com/SidOfc/mkdx/blob/2c19bb9fa9f049ee2b9e115bde097696a58b2c02/ftplugin/markdown/mkdx.vim#L131-L142). The only exception to this is enter, O and shift-O key mappings, they are [bound separately](https://github.com/SidOfc/mkdx/blob/2c19bb9fa9f049ee2b9e115bde097696a58b2c02/ftplugin/markdown/mkdx.vim#L118-L129) since they behave different from the other mappings.

I put my "init" stuff in the `ftplugin` folder and most of the beefy code in the `autoload` folder to allow for some lazy loading to happen, there is also an additional `syntax` and `after` folder but they contain nothing that has to do with the menu's. As a matter of fact, the only notion of any menu's is in `ftplugin/markdown/mkdx.vim`.

Whenever I load a markdown file in MacVim, a menu "Plugin" automatically gets appended and an entry "mkdx" exists with all the options. The same happens in terminal vim (neovim in my case, regular works too). I can view it there with `:h Plugin.mkdx` and I get the expected output.

I'm not really sure what you're trying to do with `:breakadd` in order to recreate menu's though?
Can you explain what you are expecting and what doesn't happen?

Kind regards,
Sidney Liebrand
